dev-lang/rust: Support building a BPF target sysroot

vadorovsky opened the pull request:
Rust sysroot is a directory which contains the prebuilt standard libraries (in form of `*.rlib` files) and utility binaries (e.g. `rust-objcopy`). Currently, the `dev-lang/rust` builds a sysroot only for:

* the host target
* `wasm32-unknown-unknown` target, if the `wasm` USE flag is enabled

This pull request adds a possibility to build sysroots for BPF targets (`bpfeb-unknown-none`, `bpfel-unknown-none`).

To avoid adding more USE flags and, this PR adds a new `USE_EXPAND` variable, called `RUST_SYSROOTS`, This variable is limited to bare metal targets. Sysroot for targets not falling into that category contain a local copy of compiler runtime,
libc and unwinder, so crossdev is a better fit for them.

In this PR, both `BPF` and `WebAssembly` are supported as possible `RUST_SYSROOTS` values.

See individual commits for details.

<!-- Please put the pull request description above -->

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

Kangie commented:
I like the concept here in terms of enabling additional user choice and not defaulting to building a 'fatter' package.

QA issues first:

- You will need to add a `profiles/desc/` entry for the new USE_EXPAND (and socialise it via the mailing list, you can reference this PR and direct feedback here).

https://devmanual.gentoo.org/general-concepts/use-flags/index.html#use-flag-descriptions
https://github.com/gentoo/gentoo/blob/master/profiles/desc/curl_ssl.desc
https://public-inbox.gentoo.org/gentoo-dev/20240621131552.526393-1-kangie@gentoo.org/

-  make all of the USE_EXPAND variables lowercase to be consistent (i.e. `rust_sysroots_bpf`, `rust_sysroots_wasm`). 
   + retain the shorter `wasm` rather than expanding to `webassembly` unless you really think that it's ambiguous.

- I think that `RUST_SYSROOT` works better for the USE_EXPAND name (i.e. `rust_sysroot_bpf`, `rust_sysroot_wasm`)

- I note that there's a QA issue (or three) to be addressed from the bot. These should be handled by adding the desc entry.

If we're going to all of this effort to add the USE_EXPAND:

- Is there a full list of possible sysroot values somewhere upstream? If so it should be referenced in the desc _and_ we should consider enabling additional (sane) targets now.
- do we expect new values to be added in the future?
- are there any interactions with / could this replace or supplement the existing rust-std package, which basically just builds and installs a rust sysroot?
- somewhat related to the above, should the `USE_EXPAND` contain `arch` values / could this (or something similar) be leveraged to simplify some of the existing cross-build logic rather than relying on a magic envvar?

vadorovsky commented:
> QA issues first:
> 
>  * You will need to add a `profiles/desc/` entry for the new USE_EXPAND (and socialise it via the mailing list, you can reference this PR and direct feedback here).
> 
> 
> https://devmanual.gentoo.org/general-concepts/use-flags/index.html#use-flag-descriptions https://github.com/gentoo/gentoo/blob/master/profiles/desc/curl_quic.desc https://public-inbox.gentoo.org/gentoo-dev/20240621131552.526393-1-kangie@gentoo.org/

Sorry for missing that, I will add a `profiles/desc` entry in the next push.

I proposed the change on the mailing list already:

https://archives.gentoo.org/gentoo-dev/svsyiq.2flfpudyklk6m@disroot.org/T/#u

But I guess I also need to send the patchset there?

> * make all of the USE_EXPAND variables lowercase to be consistent (i.e. `rust_sysroots_bpf`, `rust_sysroots_wasm`).
>       
>   * retain the shorter `wasm` rather than expanding to `webassembly` unless you really think that it's ambiguous.

I was thinking about keeping compatibility with `LLVM_TARGETS` (which has `BPF` and `WebAssembly`), but if you don't like the idea, I can make it lower-case. Perhaps a good move if we want to want to support cranelift or GCC as Rust compiler backends in the future.

> * I think that `RUST_SYSROOT` works better for the USE_EXPAND name (i.e. `rust_sysroot_bpf`, `rust_sysroot_wasm`)

:+1: 

> If we're going to all of this effort to add the USE_EXPAND:
> 
>  * Is there a full list of possible sysroot values somewhere upstream? If so it should be referenced in the desc _and_ we should consider enabling additional (sane) targets now.
> 
>  * do we expect new values to be added in the future?

There is a list of all supported targets here:

https://doc.rust-lang.org/nightly/rustc/platform-support.html

However, that list doesn't make it clear which targets are bare metal and which are not, without going into details. My idea would be to support only bare metal targets. By bare metal I mean both the hardware targets, like:

* `riscv32*-unknown-none-elf`
* `riscv64*-unknown-none-elf`
* `thumbv6m-none-eabi`
* `thumbv7m-none-eabi`
* `m68k-unknown-none-elf`
etc.

GPU targets, like:

* `amdgcn-amd-amdhsa`

and the VM targets, like:

* BPF
  * `bpfeb-unknown-none`
  * `bpfel-unknown-none`
* WASM
  * `wasm32-unknown-unknown`
  * `wasm64-unknown-unknown`
* other more exotic VMs
  * `riscv32im-risc0-zkvm-elf`

I would be happy to add them in separate PRs if the idea itself gets accepted. I'm starting with BPF targets, because they are blocking the update mitmproxy, which recently started using [Aya](https://github.com/aya-rs/aya) and I also care about them personally the most.

Targets which are not bare metal (and require compiler runtime + libc) should be always built with crossdev IMO, since it's reasonable to have the whole OS sysroot for them.

>     * are there any interactions with / could this replace or supplement the existing rust-std package, which basically just builds and installs a rust sysroot?

I think we should keep the rust-std package for non-bare-metal targets and keep restricting its use only to crossdev. To be specific, I would keep it primarily for all `*-*-{gnu,gnueabi,musl,musleabi}` targets.

> 
>     * somewhat related to the above, should the `USE_EXPAND` contain `arch` values / could this (or something similar) be leveraged to simplify some of the existing cross-build logic rather than relying on a magic envvar?

The reason I'm proposing this variable for bare-metal targets is that they don't need the whole OS sysroot, so I feel like requiring crossdev for them would be an overkill.

Also, given that we want to package some projects using BPF (again, mitmproxy) or WASM, we want people to be able to install such software on their regular Gentoo hosts, and we can expect amount of such software to grow, I think we shouldn't require using crossdev for building them. The ultimate goal of BPF is to be the standard for writing firewalls, CNIs, security monitoring tools etc. Maybe for now it's limited mostly to Kubernetes and advanced container usage. But imagine a future where some Rust-BPF based replacement of firewalld pops out. Requiring people to use crossdev for building such project would be quite annoying.

vadorovsky commented:
@Kangie I rebased the branch and addressed all your remarks except

> I think that RUST_SYSROOT works better for the USE_EXPAND name (i.e. rust_sysroot_bpf, rust_sysroot_wasm)

I briefly looked at all other USE_EXPAND variables and it seems like they are mostly using the plural form:

```
 ls profiles/desc/
╭────┬───────────────────────────────────────────┬──────┬─────────┬───────────────╮
│  # │                   name                    │ type │  size   │   modified    │
├────┼───────────────────────────────────────────┼──────┼─────────┼───────────────┤
│  0 │ profiles/desc/abi_mips.desc               │ file │   347 B │ a month ago   │
│  1 │ profiles/desc/abi_s390.desc               │ file │   320 B │ a month ago   │
│  2 │ profiles/desc/abi_x86.desc                │ file │   342 B │ a month ago   │
│  3 │ profiles/desc/ada_target.desc             │ file │   261 B │ 3 weeks ago   │
│  4 │ profiles/desc/alsa_cards.desc             │ file │  1.4 kB │ a month ago   │
│  5 │ profiles/desc/amdgpu_targets.desc         │ file │  2.5 kB │ a month ago   │
│  6 │ profiles/desc/apache2_modules.desc        │ file │  7.5 kB │ 2 weeks ago   │
│  7 │ profiles/desc/apache2_mpms.desc           │ file │   407 B │ a month ago   │
│  8 │ profiles/desc/calligra_features.desc      │ file │   447 B │ a month ago   │
│  9 │ profiles/desc/cameras.desc                │ file │  3.4 kB │ a month ago   │
│ 10 │ profiles/desc/collectd_plugins.desc       │ file │ 14.9 kB │ a month ago   │
│ 11 │ profiles/desc/cpu_flags_arm.desc          │ file │  1.2 kB │ a month ago   │
│ 12 │ profiles/desc/cpu_flags_ppc.desc          │ file │   375 B │ a month ago   │
│ 13 │ profiles/desc/cpu_flags_x86.desc          │ file │  2.9 kB │ a month ago   │
│ 14 │ profiles/desc/curl_quic.desc              │ file │   230 B │ a month ago   │
│ 15 │ profiles/desc/curl_ssl.desc               │ file │   272 B │ a month ago   │
│ 16 │ profiles/desc/elibc.desc                  │ file │   645 B │ a month ago   │
│ 17 │ profiles/desc/fftools.desc                │ file │  1.0 kB │ a month ago   │
│ 18 │ profiles/desc/gpsd_protocols.desc         │ file │  1.1 kB │ a month ago   │
│ 19 │ profiles/desc/grub_platforms.desc         │ file │   734 B │ a month ago   │
│ 20 │ profiles/desc/guile_single_target.desc    │ file │   284 B │ a month ago   │
│ 21 │ profiles/desc/guile_targets.desc          │ file │   232 B │ a month ago   │
│ 22 │ profiles/desc/input_devices.desc          │ file │  2.9 kB │ a month ago   │
│ 23 │ profiles/desc/kernel.desc                 │ file │   466 B │ a month ago   │
│ 24 │ profiles/desc/l10n.desc                   │ file │  4.4 kB │ a month ago   │
│ 25 │ profiles/desc/lcd_devices.desc            │ file │  7.6 kB │ a month ago   │
│ 26 │ profiles/desc/libreoffice_extensions.desc │ file │   537 B │ a month ago   │
│ 27 │ profiles/desc/llvm_slot.desc              │ file │   232 B │ a month ago   │
│ 28 │ profiles/desc/llvm_targets.desc           │ file │  1.1 kB │ a month ago   │
│ 29 │ profiles/desc/lua_single_target.desc      │ file │   307 B │ a month ago   │
│ 30 │ profiles/desc/lua_targets.desc            │ file │   285 B │ a month ago   │
│ 31 │ profiles/desc/nginx_modules_http.desc     │ file │  6.6 kB │ a month ago   │
│ 32 │ profiles/desc/nginx_modules_mail.desc     │ file │   350 B │ a month ago   │
│ 33 │ profiles/desc/nginx_modules_stream.desc   │ file │  1.7 kB │ a month ago   │
│ 34 │ profiles/desc/office_implementation.desc  │ file │   250 B │ a month ago   │
│ 35 │ profiles/desc/openmpi_fabrics.desc        │ file │   295 B │ a month ago   │
│ 36 │ profiles/desc/openmpi_ofed_features.desc  │ file │   449 B │ a month ago   │
│ 37 │ profiles/desc/openmpi_rm.desc             │ file │   256 B │ a month ago   │
│ 38 │ profiles/desc/perl_features.desc          │ file │   450 B │ a month ago   │
│ 39 │ profiles/desc/php_targets.desc            │ file │   300 B │ a month ago   │
│ 40 │ profiles/desc/postgres_targets.desc       │ file │   424 B │ a month ago   │
│ 41 │ profiles/desc/python_single_target.desc   │ file │   488 B │ 2 weeks ago   │
│ 42 │ profiles/desc/python_targets.desc         │ file │   454 B │ 2 weeks ago   │
│ 43 │ profiles/desc/qemu_softmmu_targets.desc   │ file │  1.2 kB │ a month ago   │
│ 44 │ profiles/desc/qemu_user_targets.desc      │ file │  1.6 kB │ a month ago   │
│ 45 │ profiles/desc/ruby_targets.desc           │ file │   349 B │ a month ago   │
│ 46 │ profiles/desc/rust_sysroots.desc          │ file │   168 B │ 4 minutes ago │
│ 47 │ profiles/desc/sane_backends.desc          │ file │  5.9 kB │ a month ago   │
│ 48 │ profiles/desc/uwsgi_plugins.desc          │ file │  5.1 kB │ a month ago   │
│ 49 │ profiles/desc/video_cards.desc            │ file │  3.2 kB │ 2 weeks ago   │
│ 50 │ profiles/desc/voicemail_storage.desc      │ file │   416 B │ a month ago   │
│ 51 │ profiles/desc/xtables_addons.desc         │ file │  2.1 kB │ a month ago   │
├────┼───────────────────────────────────────────┼──────┼─────────┼───────────────┤
│  # │                   name                    │ type │  size   │   modified    │
╰────┴───────────────────────────────────────────┴──────┴─────────┴───────────────╯
```

Especially the `*_FEATURES` and `*_TARGETS` variables. So I think that `RUST_SYSROOTS` is more consistent with them.

vadorovsky commented on profiles/desc/rust_sysroots.desc:
```diff
@@ -0,0 +1,5 @@
+# Copyright 1999-2025 Gentoo Authors.
```
Done

vadorovsky commented:
@mgorny @Kangie I fixed it, it works also for Rust 1.88.0 and produces a correct sysroot:

```
>>> /usr/lib/rust/1.88.0/lib/rustlib/bpfel-unknown-none/
>>> /usr/lib/rust/1.88.0/lib/rustlib/bpfel-unknown-none/lib/
>>> /usr/lib/rust/1.88.0/lib/rustlib/bpfel-unknown-none/lib/libcore-f1f74068811ae06b.rlib
>>> /usr/lib/rust/1.88.0/lib/rustlib/bpfel-unknown-none/lib/libcompiler_builtins-c799abf5d063ce62.rlib
>>> /usr/lib/rust/1.88.0/lib/rustlib/bpfel-unknown-none/lib/liballoc-7c553e4fb8184643.rlib
>>> /usr/lib/rust/1.88.0/lib/rustlib/bpfel-unknown-none/lib/librustc_std_workspace_core-cf15e4fcfd49cf5d.rlib
```

@mgorny To be able to run bpf-linker tests in ebuilds, I still need to do two things

1. Fix bpf-linker to make the sysroot download optional - I got it working, I just need to clean up the code, I will submit a PR at latest tomorrow and ping you there.
2. Package https://github.com/anakryiko/btfdump for the BTF tests

vadorovsky commented:
@mgorny @Kangie ping.

I couldn't do a revbump of 1.89.0, because it requires creating a branch in https://gitweb.gentoo.org/proj/rust-patches.git/, which I have no rights to do.

mgorny reviewed (approved):
Not exactly a Rust expert here, but it seems to build fine, and I can `cargo test --profile=dev --no-default-features --features=llvm-21` bpf-linker with it, so I guess it works as advertised!

That said, I think you should do it on 1.90.0 and skip 1.89.0 now.

arthurzam commented:
@Kangie is currently working on a patch (-r1) for 1.90, so I guess he should cherry-pick it into it

vadorovsky commented:
I'm afk today, but I could rebase tomorrow if needed. Thanks for approval!
